core-data/index.js

```diff
@@ -0,0 +1,38 @@
+/**
+ * WordPress Dependencies
+ */
+import {
+	registerReducer,
+	registerSelectors,
+	registerResolvers,
+} from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import reducer from './reducer';
+import { getCategories } from './selectors';
+
+/**
+ * Module Constants
+ */
+const MODULE_KEY = 'core';
+
+const store = registerReducer( MODULE_KEY, reducer );
+
+registerSelectors( MODULE_KEY, { getCategories } );
+registerResolvers( MODULE_KEY, {
+	getCategories: {
+		isFulfilled: ( state ) => state !== undefined,
+		fulfill: () => {
+			wp.apiRequest( { path: '/wp/v2/categories' } ).then( categories => {
```

Comment:
I'm curious how this might fit into the larger story of reducing these dependencies to the API directly, where fulfillment is merely a set of signals, or could otherwise be swapped with a specific implementation.

- Could they return actions which are observed and answered by the implementing interface?
- Could an implementing interface easily replace these resolvers with their own if they so choose?

Reply:
So the question is whether this `core-data` module I'm adding here is considered the data layer of Gutenberg in which case it should rely on swappable API layer or Should we consider this module as the adapter it-self in which case it should be entirely replaced by another implementation. Think `my-custom-data` module replacing `core-data` when Gutenberg is used in other contexts.

Honestly, I don't know the answer here, I'm still on the fence.

Reply:
I see it as the later, this module as the adapter with the whole thing being replaced.
So I could create my own data module which defines `getCategories()` and does whatever it needs and returns the data Gutenberg needs. 

I think another API layer on top assumes too much on where the data might be coming from.